service_app.rb

```diff
@@ -1,37 +1,33 @@
 require 'cgi'
 require 'yaml/store'
 require 'rack'
+require 'sqlite3'
 
 class ServiceApp
   Mirth = -> environment {
     request = Rack::Request.new(environment)
     response = Rack::Response.new
-    store = YAML::Store.new("daily_data.yml")
+    database = SQLite3::Database.new("mirth.sqlite3")
 
     if request.get? && request.path == "/show/data"
       response.write("<ul>\n")
       response.content_type = "text/html; charset=UTF-8"
 
-      all_data = {}
-      store.transaction do
-        all_data = store[:all_data]
-      end
-
+      all_data = database.execute("SELECT * FROM daily_data")
+      
       all_data.each do |daily_data|
-        response.write "<li> On this day <b>#{CGI.escapeHTML(daily_data[:date])}</b>, #{CGI.escapeHTML(daily_data[:step_count])}, #{CGI.escapeHTML(daily_data[:notes])}</li>\n"
+        response.write "<li> On this day <b>#{CGI.escapeHTML(daily_data[0])}</b>, #{CGI.escapeHTML(daily_data[1].to_s)}, #{CGI.escapeHTML(daily_data[2])}</li>\n"
       end
 
       response.write "</ul>\n"
       response.write ServiceApp.daily_data_form
 
     elsif request.post? && request.path == "/add/data"
-      new_daily_data = request.params.transform_keys(&:to_sym)
-      store.transaction do
-        store[:all_data] << new_daily_data
-      end
-
+      new_daily_data = request.params
+      query = "INSERT INTO daily_data (date, step_count, notes) VALUES (?, ?, ?)"
+      
+      database.execute(query, [new_daily_data['date'], new_daily_data['step_count'].to_i, new_daily_data['notes']])
```

Comment:
You haven’t provided any instructions for how I should create the `daily_data` table, so I don’t know what the types of these columns are, but I assume you made `step_count` an `INT` column, hence the translations with `#to_i` here and `#to_s` earlier?

If so then I guess you noticed that the `sqlite3` gem has a built-in [type translator](https://github.com/sparklemotion/sqlite3-ruby/blob/v1.4.2/lib/sqlite3/translator.rb) which “helps” the user by automatically converting a string from the database (all raw results arrive as strings) into an `Integer`, `Float`, boolean, `Date` etc object depending on the [type](https://www.sqlite.org/datatype3.html) of the column it came from. This is usually what we want, although in your case it means that you actually have to undo that work so that you can work with strings again!

(You could have just made `step_count` a `TEXT` column to avoid this, but perhaps that doesn’t feel right either.)

Reply:
> but I assume you made step_count an INT column, hence the translations with #to_i here and #to_s earlier?

Yep! 

I just made step_count an INT because I thought it was best practice to make set the column type to what the object is. In this case, we are not computing anything with step_count but just displaying it, so perhaps it's best just as a string.

Reply:
I think you probably did the right thing. Even though you’re not using the numeric value _right now_, it seems likely that in the (imaginary) future you’d want to do stuff like add up all the step counts to get a total or whatever, so having the database treat it as a number is sensible even if it involves a little more work in your Ruby code.